configuration/configuration.go

```diff
@@ -233,6 +233,12 @@ type Configuration struct {
 				Deny []string `yaml:"deny,omitempty"`
 			} `yaml:"urls,omitempty"`
 		} `yaml:"manifests,omitempty"`
+		// Indexes configures index validation
+		Indexes struct {
+			// AllowSparse allows indexes to point to manifests that
+			// do not exist in the registry
+			Sparse bool `yaml:"sparse,omitempty"`
```

Comment:
Would it be worth having a way for the manifestlist itself to indicate which images are not actually present, so that smart clients can avoid attempting to pull and getting an unexpected failure?

This metadata would need to be excluded from the manifestlist SHA generation, of course, to ensure that the manifestlist SHA remained the same regardless of which images were actually present in a given registry.

Reply:
> This metadata would need to be excluded from the manifestlist SHA generation, of course

That's the issue. Also I think to support something like that we'd really need an OCI spec change, which I imagine to be a much lengthier process, unless we just put it in as annotations, but that's a bit kludgey (and still doesn't solve the SHA issue). I also think the implementation would be better done by the registry at pull time than by the client at push time.

Clients getting a failed pull is likely to look (to the end user) the same as if the index wasn't there at all, although hopefully with a decent error message.

Reply:
> Clients getting a failed pull is likely to look (to the end user) the same as if the index wasn't there at all, although hopefully with a decent error message.

right, but it will be hard(impossible for the client?) to disambiguate "this wasn't found and that's to be expected because this is a sparse manifestlist" from "this wasn't found, maybe you don't have access to that image?  or the registry is corrupt?  or something else strange happened?", for the client/user.

Reply:
I don't think it would be hard to disambiguate:
- If the index is found but the reference 404's, its a sparse index
- If you don't have access, you'd get a 403
- If the registry is corrupt, you'd get a 500 (or whatever you'd get if a standard image was corrupt)

Once the client is redirected to the referenced manifest, it's standard processing. I don't think any client will break in odd ways, but I do think that clients will have the opportunity to improve their error messages :)

Reply:
> If the index is found but the reference 404's, its a sparse index
> If the registry is corrupt, you'd get a 500 (or whatever you'd get if a standard image was corrupt)

i meant that the missing image itself could be deliberate(sparse) or an indication of a corrupt registry.  The registry itself would not even know whether it should "expect" to have found the image in question or not.  (e.g. imagine a pruning process that goes wrong)

> If you don't have access, you'd get a 403

pretty sure i have seen implementations that prefer to give you a 404 rather than acknowledge the image content exists, but you don't have access to it, but i could be wrong.

Reply:
If the image is missing, it's missing. Does it matter if that's because it's a sparse index targetting a different architecture, or if the registry is corrupt?

The same applies to any missing image - is it missing because it wasn't put there or because the registry lost it?

To come back around to the PR - are you suggesting that sparse indexes don't work unless there is a way for the client to be informed about what exists, or just saying that it would be better if that option also existed? I agree that a way for the registry to tell the client what to expect would be an improvement :)

Reply:
> To come back around to the PR - are you suggesting that sparse indexes don't work unless there is a way for the client to be informed about what exists, or just saying that it would be better if that option also existed? 

the latter.  I was trying to make the case for why explicit metadata in the manifestlist that says "this image isn't actually available" would be useful.

Reply:
Then we do agree :)

I feel like for that to happen, we'd need to extend the manifest to have a section called "registry-metadata" or similar, for information that is specific to that registry that could be explicitly removed from the sha generation.

Alternatively, put it in the http header where it wouldn't count towards the digest anyway, but that would still need to be specced somewhere.

Not sure I have the appetite to suggest the spec change required though.